Advanced-Recruit-Training/src/LogFileReader.cpp

```diff
@@ -37,34 +36,55 @@ bool LogFileReader::readAll(const QString& fileName)
         if (!parseLine(line, batteryData))
         {
             qDebug() << "Error while parsing" << line;
-            // return false;
+             return false;
         }
         else
         {
-            // This is how to send out a signal in QT using the emit keyword.
-            // This line notifies the classes listening to this signal
-            // that battery data has been received.
             emit batteryDataReceived(batteryData);
         }
     }
 
     return true;
 }
 
-/* File input is a csv file in the format of hh:mm:ss:zzz, voltage, current
- * Need to implement error checking for the correct number of values and
- * that the conversion from string to double is sucessful.*/
+
 bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) const
 {
     QStringList sections = line.split(BATDATA_DELIMITER);
 
+    if (sections.length() < COLUMNS)
+    {
+        qDebug() << "sections length is less than 3";
+        return false;
+    }
+
+    else if (sections.length() > COLUMNS)
+    {
+        qDebug() << "sections length is greater than 3";
+        return false;
+    }
+
     QString timeString = sections.at(0);
     batteryData.time = QTime::fromString(timeString, STRING_TIME_FORMAT);
 
-    batteryData.voltage = sections.at(1).toDouble();
+    bool doubleConversionVoltage;
+    bool doubleConversionCurrent;
 
-    batteryData.current = sections.at(2).toDouble();
+    batteryData.voltage = sections.at(1).toDouble(&doubleConversionVoltage);
 
-    return true;
+    batteryData.current = sections.at(2).toDouble(&doubleConversionCurrent);
+
+    if (doubleConversionVoltage == false || doubleConversionCurrent == false)
```

Comment:
I think I mentioned this before, but try formatting this like you did for `if(batteryData.time.isValid())` below.